demos/pkey/EVP_PKEY_RSA_keygen.c

```diff
@@ -0,0 +1,279 @@
+#include <openssl/pem.h>
+
+/* A property query used for selecting algorithm implementations. */
+static const char *propq = NULL;
+
+/*
+ * Generates an RSA public-private key pair and returns it.
+ * The number of bits is specified by the bits argument.
+ *
+ * This uses the long way of generating an RSA key.
+ */
+static EVP_PKEY *generate_rsa_key_long(OSSL_LIB_CTX *libctx, unsigned int bits)
+{
+    EVP_PKEY_CTX *genctx = NULL;
+    EVP_PKEY *pkey = NULL;
+    unsigned int primes = 2;
+
+    /* Create context using RSA algorithm. "RSA-PSS" could also be used here. */
+    genctx = EVP_PKEY_CTX_new_from_name(libctx, "RSA", propq);
+    if (genctx == NULL) {
+        fprintf(stderr, "EVP_PKEY_CTX_new_from_name() failed\n");
+        goto cleanup;
+    }
+
+    /* Initialize context for key generation purposes. */
+    if (EVP_PKEY_keygen_init(genctx) <= 0) {
+        fprintf(stderr, "EVP_PKEY_keygen_init() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * Here we set the number of bits to use in the RSA key.
+     * See comment at top of file for information on appropriate values.
+     */
+    if (EVP_PKEY_CTX_set_rsa_keygen_bits(genctx, bits) <= 0) {
+        fprintf(stderr, "EVP_PKEY_CTX_set_rsa_keygen_bits() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * It is possible to create an RSA key using more than two primes.
+     * Do not do this unless you know why you need this.
+     * You ordinarily do not need to specify this, as the default is two.
+     *
+     * Both of these parameters can also be set via EVP_PKEY_CTX_set_params, but
+     * these functions provide a more concise way to do so.
+     */
+    if (EVP_PKEY_CTX_set_rsa_keygen_primes(genctx, primes) <= 0) {
+        fprintf(stderr, "EVP_PKEY_CTX_set_rsa_keygen_primes() failed\n");
+        goto cleanup;
+    }
+
+    /*
+     * Generating an RSA key with a number of bits large enough to be secure for
+     * modern applications can take a fairly substantial amount of time (e.g.
+     * one second). If you require fast key generation, consider using an EC key
+     * instead.
+     *
+     * If you require progress information during the key generation process,
+     * you can set a progress callback using EVP_PKEY_set_cb; see the example in
+     * EVP_PKEY_generate(3).
+     */
+    fprintf(stderr, "Generating RSA key, this may take some time...\n");
+    if (EVP_PKEY_generate(genctx, &pkey) <= 0) {
+        fprintf(stderr, "EVP_PKEY_generate() failed\n");
+        goto cleanup;
+    }
+
+    /* pkey is now set to an object representing the generated key pair. */
+
+cleanup:
+    EVP_PKEY_CTX_free(genctx);
+    return pkey;
+}
+
+/*
+ * Generates an RSA public-private key pair and returns it.
+ * The number of bits is specified by the bits argument.
+ *
+ * This uses a more concise way of generating an RSA key, which is suitable for
+ * simple cases. It is shown for demonstration but is not used by the code
+ * below.
+ */
+#if 0
```

Comment:
I wonder if it would be better to not "if 0" this out? Just call it (e.g. maybe depending on a command line option or something) Since other discussions have said that we would like to build the demos (at least in CI) I'd rather avoid "if 0" since that sort of code has a tendency to rot over time.

Reply:
I agree with @mattcaswell here, lets include the code and have an alternate path for call it.

Reply:
@hlandau would you please consider this? Or maybe in an additional PR as this one could get merged now.

Reply:
Done.